### DISCUSSION TITLE
Odd looking placement density in designs/nangate45/gcd

### DISCUSSION THREAD

**Discussion by oharboe** [The-OpenROAD-Project/OpenROAD#3953] [General]
Run `make designs/nangate45/gcd/config.mk &&  make designs/nangate45/gcd/config.mk gui_final` and enable placement heat map.

This looks odd. There is less going on in the top and right edges, but it still "looks wrong" to me.

Thoughts?

![image](https://github.com/The-OpenROAD-Project/OpenROAD/assets/2798822/532929ca-34bd-4bce-b3b2-df547ff91e26)

**Comment by maliberty** [comment]
It does look a bit odd.  @AcKoucher can you take a look?

**Reply by AcKoucher** [reply]
Yes. I'll take a look.

**Comment by AcKoucher** [comment]
With nets being draw it's a bit hard to see. I just ran master in nangate45/gcd and as far as I see, everything looks ok.
![image](https://github.com/The-OpenROAD-Project/OpenROAD/assets/104802710/28693f96-c91d-4b92-a1af-4985741156c5)

**Reply by oharboe** [reply]
What version are you running. I seem to recall that the black stripe to the top and right was fixed?

**Reply by AcKoucher** [reply]
Oh, I see your point. I am running master. The "black stripe to the top and right" was indeed fixed. I'll do my best to explain it below.

The thing is the last row and the last column of the grid can be smaller than the others. Consider here that the map extends from the bottom-left corner to the up-right corner.
Before the fix was made (PR #3876), the last row and the last column were not being populated with the correct information.

After the fix (as we can see in my screenshot), we populate those correctly:
What we see in my screenshot is the case where the last row and the last column are black, because there is simply no instance in the area that they cover, which is the correct behavior even though they are thinner. If there was, in the area that the last row and last column cover, some instance to make the density>0, that area would have its color changed to something different than black.

The most relevant thing here is that the map is black only in the areas where density == 0;

**Comment by AcKoucher** [comment]
I just realized in the discussion name it says Placement Density, but in your image we see Routing Congestion.

**Reply by oharboe** [reply]
Includes version number and the menu so you can see the heatmap selection:

![image](https://github.com/The-OpenROAD-Project/OpenROAD/assets/2798822/9951eac1-d83a-4201-98fc-c62e427daa26)

![image](https://github.com/The-OpenROAD-Project/OpenROAD/assets/2798822/8a695039-c5ea-406b-ad5f-e7827a06372c)

**Reply by oharboe** [reply]
![image](https://github.com/The-OpenROAD-Project/OpenROAD/assets/2798822/3fb93ff1-0ede-48af-93b5-c026534a2732)
![image](https://github.com/The-OpenROAD-Project/OpenROAD/assets/2798822/c75ad1a8-1c32-40ba-9d6f-8facce148700)

**Reply by AcKoucher** [reply]
Regarding placement, I've answered above.

As for the Routing Congestion, I talked to @eder-matheus and indeed it seems like there might be a problem with the data that GRT is sending the GUI. He will look into it so we can understand it better.
